src/Core/FlatpakBackend.vala

```diff
@@ -0,0 +1,860 @@
+/*-
+ * Copyright 2019 elementary, Inc. (https://elementary.io)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored by: David Hewitt <davidmhewitt@gmail.com>
+ */
+
+public class AppCenterCore.FlatpakBackend : Backend, Object {
+    // AppStream data has to be 1 hour old before it's refreshed
+    public const uint MAX_APPSTREAM_AGE = 3600;
+
+    private AsyncQueue<Job> jobs = new AsyncQueue<Job> ();
+    private Thread<bool> worker_thread;
+
+    private Gee.HashMap<string, Package> package_list;
+    private AppStream.Pool appstream_pool;
+
+    // This is OK as we're only using a single thread
+    // This would have to be done differently if there were multiple workers in the pool
+    private bool thread_should_run = true;
+
+    public bool working { get; private set; }
+
+    private string local_metadata_path;
+
+    private bool worker_func () {
+        while (thread_should_run) {
+            working = false;
+            var job = jobs.pop ();
+            working = true;
+            switch (job.operation) {
+                case Job.Type.REFRESH_CACHE:
+                    refresh_cache_internal (job);
+                    break;
+                case Job.Type.INSTALL_PACKAGE:
+                    install_package_internal (job);
+                    break;
+                case Job.Type.UPDATE_PACKAGE:
+                    update_package_internal (job);
+                    break;
+                case Job.Type.REMOVE_PACKAGE:
+                    remove_package_internal (job);
+                    break;
+                default:
+                    assert_not_reached ();
+            }
+        }
+
+        return true;
+    }
+
+    private FlatpakBackend () {
+        worker_thread = new Thread<bool> ("flatpak-worker", worker_func);
+        appstream_pool = new AppStream.Pool ();
```

Comment:
Is there a way we could make this pool shared across backends? I'm not sure if creating the pool causes disk reads or not but it would be nice.

Reply:
I would prefer to keep these separate so that there can be no conflicts of AppStream data between Flatpak and PackageKit. I know it's been designed to allow multiple packages with the same ID to come from different sources, but I'd had to do some custom metadata paths, XML wrangling and icon symlinking to get the Flatpak metadata to work properly, and I'd prefer to keep that separate.